### PR TITLE
Add BHASA scenarios, run specs, and metrics

### PR DESCRIPTION
**Pull request by raileymontalan**

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -110,6 +110,11 @@ mongo =
 unitxt =
     evaluate~=0.4.1
 
+bhasa = 
+    pythainlp==5.0.0
+    pyonmttok==1.37.0
+    sacrebleu~=2.2.1
+
```
Comment (weiqipedia) [inline setup.cfg:117]: I think we might need to add `cfrm-helm[bhasa]` at line 265 under "all" so that these will be installed when the user chooses to install everything

File: src/helm/benchmark/metrics/bhasa_metrics.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
+import re, string
+from dataclasses import replace
+from typing import Callable, List, Dict, Any, cast
+from functools import partial
+
+from helm.common.hierarchical_logger import hlog
```
Comment (weiqipedia) [inline src/helm/benchmark/metrics/bhasa_metrics.py]: This is not used

File: src/helm/benchmark/metrics/bhasa_metrics.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+from helm.benchmark.adaptation.scenario_state import ScenarioState
+from helm.benchmark.adaptation.request_state import RequestState
+from helm.benchmark.adaptation.adapter_spec import AdapterSpec
+from helm.benchmark.scenarios.scenario import Reference
```
Comment (weiqipedia) [inline src/helm/benchmark/metrics/bhasa_metrics.py]: This is also not used

File: src/helm/benchmark/metrics/xlsum/scoring.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    Returns:
+      A 2-d matrix of (bounds, measure). There are three bounds: low (row 0),
+      mid (row 1) and high (row 2). Mid is always the mean, while low and high
+      bounds are specified by self._confidence_interval (which defaults to 0.95
+      meaning it will return the 2.5th and 97.5th percentiles for a 95%
+      confidence interval on the mean).
+    """
+
+    # Matrix of (bootstrap sample, measure).
+    sample_mean = np.zeros((self._n_samples, matrix.shape[1]))
```
Comment (weiqipedia) [inline src/helm/benchmark/metrics/xlsum/scoring.py]: Shouldn't there be a `np.random.seed(4156)` after this line to ensure reproducible bootstrap aggregation?

File: src/helm/benchmark/run_specs/bhasa_run_specs.py [hunk 5]
```diff
@@ -0,0 +1,587 @@
...
+
+# NLU
+@run_spec_function("indicqa")
+def get_indicqa_spec() -> RunSpec:
+    name = "indicqa"
+
+    adapter_spec = get_generation_adapter_spec(
+        instructions="உங்களுக்கு ஒரு பத்தியும் ஒரு கேள்வியும் தரப்படும். தரப்பட்ட பத்தியிலிருந்து கேள்விக்கான பதிலைக் கண்டறியவும்.",
+        output_noun="பதில்",
+        stop_sequences=["<|endoftext|>", "\n"],
```
Comment (weiqipedia) [inline src/helm/benchmark/run_specs/bhasa_run_specs.py]: Let's remove the `<|endoftext|>` since that is specific to our model and will be fixed eventually

File: src/helm/benchmark/scenarios/bhasa_scenario.py [hunk 6]
```diff
@@ -0,0 +1,1494 @@
...
+            'train': TRAIN_SPLIT,
+            'test': TEST_SPLIT
+        }
+
+    def get_instances(self, output_path) -> List[Instance]:
+        dataset = datasets.load_dataset("ai4bharat/IndicQA", "indicqa.ta")
+        df = dataset['test'].to_pandas()
+        df = df[df["answers"].apply(lambda x: len(x["text"][0].strip()) > 0)]
+        df_test = df.sample(n=100, random_state=7900)
+        df_train = df[~df.apply(tuple,1).isin(df_test.apply(tuple,1))]
```
Comment (weiqipedia) [inline src/helm/benchmark/scenarios/bhasa_scenario.py]: This can be done with 

```python
df_train = df[~df.index.isin(df_test.index)]
```

Would be clearer in terms of what it is doing

File: src/helm/benchmark/scenarios/bhasa_scenario.py [hunk 7]
```diff
@@ -0,0 +1,1494 @@
...
+
+    def __init__(self):
+        super().__init__()
+        self.splits = {
+            'train': TRAIN_SPLIT,
+            'test': TEST_SPLIT
+        }
+
+    def get_instances(self, output_path) -> List[Instance]:
+        dataset = datasets.load_dataset("ai4bharat/IndicQA", "indicqa.ta")
```
Comment (weiqipedia) [inline src/helm/benchmark/scenarios/bhasa_scenario.py]:
```python
dataset = datasets.load_dataset("ai4bharat/IndicQA", "indicqa.ta", split="test")
df = dataset.to_pandas()
```

File: src/helm/benchmark/scenarios/bhasa_scenario.py [hunk 8]
```diff
@@ -0,0 +1,1494 @@
...
+                    input=input,
+                    references=references, 
+                    split=self.splits[split]
+                )
+                outputs.append(instance)
+        return outputs
+    
+# LD
+    
+class LINDSEAMPScenario(Scenario):
```
Comment (weiqipedia) [inline src/helm/benchmark/scenarios/bhasa_scenario.py]: I think LINDSEASyntaxScenario might be clearer

File: src/helm/benchmark/run_specs/bhasa_run_specs.py [hunk 9]
```diff
@@ -0,0 +1,625 @@
...
+# 1.1 Indonesian: TyDiQA
+@run_spec_function("tydiqa")
+def get_tydiqa_spec() -> RunSpec:
+    name = "tydiqa"
+
+    adapter_spec = get_generation_adapter_spec(
+        instructions="Anda akan diberikan sebuah paragraf dan sebuah pertanyaan. Jawablah pertanyaannya dengan mengekstrak jawaban dari paragraf tersebut.",
+        output_noun="Jawaban",
+        stop_sequences=["\n"],
+        max_tokens=128,
```
Comment (weiqipedia) [inline src/helm/benchmark/run_specs/bhasa_run_specs.py]: Currently using the following max_tokens:
1. QA: 128
2. Sentiment: 8
3. Toxicity: 8
4. Translation: 128
5. Summarization: 128
6. NLI: 8
7. Causal: 8

To review if these numbers are appropriate based on longest reference and typical tokenizer

File: src/helm/benchmark/metrics/bhasa_metrics.py [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+        self.language: str = language
+        self.rouge_fns = {
+            "rouge_l": self._get_bhasa_rouge_function("rougeL"),
+        }
+
+    def _get_bhasa_rouge_function(self, rouge_type: str) -> Callable[[str, str], float]:
+        if self.language == "th":
+            scorer = rouge_scorer.RougeScorer(
+                [rouge_type],
+                use_stemmer=True,
```
Comment (weiqipedia) [inline src/helm/benchmark/metrics/bhasa_metrics.py]: use_stemmer can be False since we're not using it right?

File: src/helm/benchmark/metrics/bhasa_metrics.py [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+                callable_tokenizer=tokenizers.ThaiTokenizer())
```
Comment (weiqipedia) [inline src/helm/benchmark/metrics/bhasa_metrics.py]: Actually there is an option to just do
```python
scorer = rouge_scorer.RougeScorer(["rougeL"], use_stemmer=False, lang="thai")
```
right?

We can do something like

```python
language = "thai" if self.language == "th" else None
scorer = rouge_scorer.RougeScorer(["rougeL"], use_stemmer=False, lang=language)
```

File: src/helm/benchmark/metrics/bhasa_metrics.py [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+        else:
+            scorer = rouge_scorer.RougeScorer([rouge_type], use_stemmer=True)
```
Comment (weiqipedia) [inline src/helm/benchmark/metrics/bhasa_metrics.py]: This is not using the bootstrap aggregator?

There should be something like
```python
aggregator = BootstrapAggregator()
for score in list_of_rouge_scores_for_every_summary:
        aggregator.add_scores(score)
aggregates = aggregator.aggregate()
mid_scores = aggregates["rougeL"].mid
```

This will allow us to get the ROUGE-L F1 mid scores

File: src/helm/benchmark/scenarios/bhasa_scenario.py [hunk 13]
```diff
@@ -0,0 +1,1570 @@
...
+
+        Paragraf: <text>
+        Pertanyaan: <question>
+        Jawaban:
+
+
+    Target completion:
+        <answer>
+
+    @inproceedings{ruder-etal-2021-xtreme,
```
Comment (weiqipedia) [inline src/helm/benchmark/scenarios/bhasa_scenario.py]: TyDiQA has a publication of its own:

https://aclanthology.org/2020.tacl-1.30/

File: src/helm/benchmark/scenarios/bhasa_scenario.py [hunk 14]
```diff
@@ -0,0 +1,1570 @@
...
+        ...
+
+        Kalimat: <text>
+        Jawaban:
+
+    Target completion:
+        <sentiment> (<sentiment>:positive or negative or neutral)
+
+    @inproceedings{van2018uit,
+        title={UIT-VSFC: Vietnamese students’ feedback corpus for sentiment analysis},
```
Comment (weiqipedia) [inline src/helm/benchmark/scenarios/bhasa_scenario.py]: This should be NusaX (https://aclanthology.org/2023.eacl-main.57/)

File: src/helm/benchmark/scenarios/bhasa_scenario.py [hunk 15]
```diff
@@ -0,0 +1,1570 @@
...
+
+    def download_dataset(self, output_path: str):
+        URLS = {
+            "test": "https://raw.githubusercontent.com/IndoNLP/nusax/main/datasets/sentiment/indonesian/test.csv",
+            "train": "https://raw.githubusercontent.com/IndoNLP/nusax/main/datasets/sentiment/indonesian/train.csv",
+        }
+
+        dataset = {}
+        for split in self.splits.keys():
+            dataset[split] = []
```
Comment (weiqipedia) [inline src/helm/benchmark/scenarios/bhasa_scenario.py]: Will be defined later anyway, no need to create an empty list here

File: src/helm/benchmark/scenarios/bhasa_scenario.py [hunk 16]
```diff
@@ -0,0 +1,1570 @@
...
+        }
+
+    def download_dataset(self, output_path: str):
+        URL = "https://github.com/PyThaiNLP/wisesight-sentiment/raw/master/huggingface/data.zip"
+        data_path = os.path.join(output_path, "data")
+        ensure_file_downloaded(source_url=URL, target_path=data_path, unpack=True)
+
+        dataset = {}
+        for split in self.splits.keys():
+            dataset[split] = []
```
Comment (weiqipedia) [inline src/helm/benchmark/scenarios/bhasa_scenario.py]: No need to define this

File: src/helm/benchmark/scenarios/bhasa_scenario.py [hunk 17]
```diff
@@ -0,0 +1,1570 @@
...
+        ...
+
+        வாக்கியம்: <text>
+        பதில்: <answer>
+
+    Target completion:
+        <sentiment> (<sentiment>:positive or negative)
+
+    @article{Doddapaneni2022towards,
+        title={Towards Leaving No Indic Language Behind: Building Monolingual Corpora, Benchmark and Models for Indic Languages},
```
Comment (weiqipedia) [inline src/helm/benchmark/scenarios/bhasa_scenario.py]: We can use the ACL version (https://aclanthology.org/2023.acl-long.693/)

File: src/helm/benchmark/scenarios/bhasa_scenario.py [hunk 18]
```diff
@@ -0,0 +1,1570 @@
...
+            for _, row in data.iterrows():
+                passage = row["passage_text"].strip()
+                question = row["question_text"].strip()
+                input = PassageQuestionInput(
+                    passage=passage,
+                    question=question,
+                    passage_prefix="Paragraf: ",
+                    question_prefix="Pertanyaan: ",
+                )
+                output = Output(text=row["answers"]["text"][0].strip())
```
Comment (weiqipedia) [inline src/helm/benchmark/scenarios/bhasa_scenario.py]: TyDiQA has more than one reference answer per question, so we need to have multiple references or something? (Instead of just taking the 0th index)

File: src/helm/benchmark/run_specs/bhasa_run_specs.py [hunk 19]
```diff
@@ -585,28 +585,22 @@ def get_xcopa_spec(language="id") -> RunSpec:
     return generate_xcopa_run_spec(language)
 
 # D. Linguistic Diagnostics
-#  1. Syntax (Minimal Pairs)
-#  2. Pragmatics
+#   1. Syntax: Minimal Pairs
+#   2. Semantics: Pragmatic Reasoning (single sentece)
```
Comment (weiqipedia) [inline src/helm/benchmark/run_specs/bhasa_run_specs.py]: This is Pragmatics, not Semantics

File: src/helm/benchmark/presentation/run_entries_bhasa.conf [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+    {description: "xcopa:model=text,language=th", priority: 1},
+    {description: "xcopa:model=text,language=vi", priority: 1},
+
+
+    ##   D. Linguistic Diagnostics (LINDSEA)
+    
+    ###  1. Syntax: Minimal Pairs
+    {description: "lindsea_syntax_minimal_pairs:model=text,language=id,max_train_instances=0", priority: 1},
+
+    ###  2. Semantics: Pragmatic Reasoning (single sentece)
```
Comment (weiqipedia) [inline src/helm/benchmark/presentation/run_entries_bhasa.conf]: Pragmatics

File: src/helm/benchmark/scenarios/bhasa_scenario.py [hunk 21]
```diff
@@ -1568,3 +1615,166 @@ def get_instances(self, output_path) -> List[Instance]:
...
+        file = "pragmatic_reasoning_single"
+        target_path_file = os.path.join(output_path, file)
+        ensure_file_downloaded(source_url=URL, target_path=target_path_file)
+        dataset = pd.read_json(target_path_file, lines=True)
+        return dataset
+
+
+    def get_instances(self, output_path) -> List[Instance]:
+        data = self.download_dataset(output_path)
+        dataset = datasets.Dataset.from_pandas(data).train_test_split(test_size=0.9, seed=5018)
```
Comment (weiqipedia) [inline src/helm/benchmark/scenarios/bhasa_scenario.py]: LINDSEA is not split into train and test sets.

File: src/helm/benchmark/run_specs/bhasa_run_specs.py [hunk 22]
```diff
@@ -590,14 +591,33 @@ def get_xcopa_spec(language="id") -> RunSpec:
...
+            empty_input=True,
+        )
+    elif mcq:
+        adapter_spec = get_multiple_choice_separate_adapter_spec(
+            method=ADAPT_MULTIPLE_CHOICE_SEPARATE_ORIGINAL,
+            empty_input=True,
+        )
+    else:
+        adapter_spec = get_generation_adapter_spec(
+            output_noun=lindsea_syntax_minimal_pairs_prompts[language]["output_noun"] + "\n" + lindsea_syntax_minimal_pairs_prompts[language]["output_noun"],
```
Comment (weiqipedia) [inline src/helm/benchmark/run_specs/bhasa_run_specs.py]: This produces
```markdown
Jawaban
Jawaban:
```

Need to include the `output_prefix`

File: src/helm/benchmark/run_specs/bhasa_run_specs.py [hunk 23]
```diff
@@ -0,0 +1,689 @@
...
+    },
+}
+
+@run_spec_function("lindsea_syntax_minimal_pairs")
+def get_lindsea_syntax_minimal_pairs_spec(language="id", mcq=False, calibrated=False) -> RunSpec:
+    name = f"lindsea_syntax_minimal_pairs_{language}"
+
+    if mcq and calibrated:
+        adapter_spec = get_multiple_choice_separate_adapter_spec(
+            method=ADAPT_MULTIPLE_CHOICE_SEPARATE_CALIBRATED,
```
Comment (weiqipedia) [inline src/helm/benchmark/run_specs/bhasa_run_specs.py]: Calibrated actually refers to normalizing by the probability of "Answer: {reference}", no need for this mode

File: src/helm/benchmark/run_specs/bhasa_run_specs.py [hunk 24]
```diff
@@ -0,0 +1,689 @@
...
+            empty_input=True,
+        )
+    elif mcq:
+        adapter_spec = get_multiple_choice_separate_adapter_spec(
```
Comment (weiqipedia) [inline src/helm/benchmark/run_specs/bhasa_run_specs.py]: Instructions should be defined here and not the Scenario to avoid the instructions appearing during MultipleChoiceSeparate

File: src/helm/benchmark/scenarios/bhasa_scenario.py [hunk 25]
```diff
@@ -0,0 +1,1894 @@
...
+                ensure_file_downloaded(source_url=URLS[split][file], target_path=target_path_file)
+                with open(target_path_file, "r") as f:
+                    lines = f.readlines()
+                    for line in lines:
+                        dataset[split][file].append(str(line).strip())
+            data = pd.DataFrame({
+                "text": dataset[split]["sentences"],
+                "label": dataset[split]["sentiments"]
+            })
+            dataset[split] = data
```
Comment (weiqipedia) [inline src/helm/benchmark/scenarios/bhasa_scenario.py]: Test set needs to be sampled
